### PR TITLE
Cnfe bundle loader

### PR DESCRIPTION
**Pull request by tjwatson**

### CODE REVIEW

**Comment by tjwatson** [comment]
#build

**Comment by tjwatson** [comment]
#build

**Comment by tjwatson** [comment]
#build

**Comment by tjwatson** [comment]
#build

**Comment by tjwatson** [comment]
#build

**Comment by tjwatson** [comment]
#build

**Comment by tjwatson** [comment]
#build

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java
Comment (jhanders34) [inline dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java:719]: Should we pass false to this loadClass as well?

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java
Comment (jhanders34) [inline dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java]: Should this just be the first check in the method and skip everything else if the parent is null?  Otherwise you end up calling findLoadedClass twice.

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java
Comment (jhanders34) [inline dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java:727]: This has been wrong for a while it would appear.  findClass also does a findLoadedClass().  It appears that it shouldn't do that and should just look for the class without looking at the loaded classses.  Of course this will require and update to the ParentLastClassLoader to call findLoadedClass() before calling findClass() if we change the way that findClass() works.

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ParentLastClassLoader.java
Comment (jhanders34) [inline dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ParentLastClassLoader.java]: Technically this could be changed to call a different method that didn't throw the exception similar to what you did in the GatewayClassLoader.

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/UnifiedClassLoader.java
Comment (jhanders34) [inline dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/UnifiedClassLoader.java]: Later in the method they do a doPriv for getParent.  Do we need to do the same here?  Of course this can be avoided by storing the parent in the class like what is done with the other implementations.

Also, there is no handling of the parent being null.

**Review by jhanders34 — CHANGES_REQUESTED** [review]

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java
Comment (tjwatson) [reply dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java:719]: I find this annoying about Java, but a subclass can only call protected methods on the super instance or other types that are in the same package (or self).

ClassLoader.loadClass(name) just ends up calling loadClass(name, false) anyway.

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java
Comment (tjwatson) [reply dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java]: We must check findLoadedClass before delegating to parent, otherwise we will continually delegate to parent for already defined classes.

findLoadedClass should be super fast, it should only be called twice if we have not already loaded class so I doubt it is measurable hit.  If we find it does then I suggest we have another variant of findClass that avoids getting the class loading lock and check for findLoadedClass because both are already done in this context.

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java
Comment (tjwatson) [reply dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java:727]: I think findClass currently must call findLoadedClass because direct calls to LibertyLoader (which AppClassLoader eventually extends) done in `AppClassLoader.findClassCommonLibraryClassLoaders(String)`

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ParentLastClassLoader.java
Comment (tjwatson) [reply dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ParentLastClassLoader.java]: I suppose this is true.  My approach here was to reuse the generated exception from findClass when the parent returned null.  If we did what you suggest then we would have to generate an exception out of the parent delegation.  At that point there is probably little value in having NoClassNotFoundLoader parent, or at least there is little value in calling parent.loadClassNoException here.  But you are correct that it would save us one CNFE for searching self before parent when the class exists in parent.

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/UnifiedClassLoader.java
Comment (tjwatson) [reply dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/UnifiedClassLoader.java]: I agree:

- need to handle null parent (although I don't think this loader will ever have a null parent.
- should store parent in an accessible field to avoid security check locally.

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java
Comment (jhanders34) [reply dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java]: Calling super will end up calling findLoadedClass first already, just like it did before you made this change.

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java
Comment (jhanders34) [reply dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java:727]: ok I will take this under advisement when looking at doing some cleanup here in a follow one PR.

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ParentLastClassLoader.java
Comment (jhanders34) [reply dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ParentLastClassLoader.java]: Yeah as you suggest, this can be done in a separate PR after this one.  An example where this would improve things is in the case that it is using liberty API / SPI or spec APIs in the application.  This is outside the scope of the PR and can be done in the follow on one.

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java
Comment (jhanders34) [reply dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java]: If you want me to clean this up in a follow on PR that is fine, but your analysis it incorrect for sure.

Comment (tjwatson) [reply dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java]: I see what you mean now.  I misunderstood you to suggest we check instance of NoClassNotFoundLoader first.  But yes we can check parent == null first here.

Comment (tjwatson) [reply dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java]: I fixed this up in a follow on commit.

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/UnifiedClassLoader.java
Comment (tjwatson) [reply dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/UnifiedClassLoader.java]: I fixed this in a follow on commit.

**Comment by tjwatson** [comment]
#libby
#build

**Review by jhanders34 — APPROVED** [review]

**Comment by tjwatson** [comment]
Fixes #19297

**Comment by tjwatson** [comment]
While I believe this PR to be ready to merge I plan to wait until after the current release (22.0.0.2) build to be finalized before merging.  At that time I will rebase and do another build to confirm.  The last build seems good, but had an untriaged issue detected in the UI tests:

```
System info: host: 'ne-f88cd-55ce-47ca-afe2-8b5e94d72c3b-ebc-2rzu65z3hhmg', ip: '127.0.1.1', os.name: 'Linux', os.arch: 'amd64', os.version: '4.15.0-20-generic', java.version: '11.0.12'
Driver info: driver.version: RemoteWebDriver
	at org.openqa.selenium.remote.http.netty.NettyHttpHandler.makeCall(NettyHttpHandler.java:72)
	at org.openqa.selenium.remote.http.netty.NettyHttpHandler$$Lambda$249/0x0000000000000000.execute(Unknown Source)
```

I believe this is not related to my changes here.  Seems to be infrastructure related for the UI tests that use selenium.

**Comment by tjwatson** [comment]
#build

**Comment by tjwatson** [comment]
#libby

**Pull request merged**
